### PR TITLE
udn: auto-exclude GW and mp0 IPs from the subnet

### PR DESCRIPTION
**Pull request by maiqueb**
This commit auto-excludes the gateway and management port IP addresses from the subnet range for user defined primary networks of `layer2` topology.

We are [already getting this behavior](https://github.com/ovn-org/ovn-kubernetes/blob/e3ad29066d00085d9a18b49751288d2fe5f3d10a/go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go#L24) on all L3 secondary networks + the default cluster network; thus we do not have to do anything user-defined primary networks.

#### What this PR does and why is it needed

This PR  auto-excludes the gateway and management port IP addresses from the
subnet range for user defined primary networks.

#### Which issue(s) this PR fixes

Fixes #

#### How to verify it

Covered by unit tests (introduced in this PR).

#### Description for the changelog

Auto-exclude GW and mp0 IPs from the subnets of user defined primary networks.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: go-controller/pkg/util/multi_network.go [hunk 1]
```diff
@@ -432,6 +432,10 @@ func newLayer2NetConfInfo(netconf *ovncnitypes.NetConf) (NetInfo, error) {
 		return nil, fmt.Errorf("invalid %s netconf %s: %v", netconf.Topology, netconf.Name, err)
 	}
 
+	if netconf.Role == types.NetworkRolePrimary {
+		excludes = primaryNetworkExcludedIPs(subnets)
```
Comment (qinqon) [inline go-controller/pkg/util/multi_network.go:436]: This would override users configuring the NAD exclude subnet field.

**Review by qinqon — CHANGES_REQUESTED** [review]

File: go-controller/pkg/util/multi_network.go (hunk 1 above)
Comment (maiqueb) [reply go-controller/pkg/util/multi_network.go:436]: I know. I assumed we don't want users to use that for user defined primary nets. 

... do we ? @tssurya

Comment (tssurya) [reply go-controller/pkg/util/multi_network.go:436]: hmm good question I think we should continue to support excludesubnets just to stay consistent with user expectations when using a NAD. maybe we should append to user input if any.

**Comment by tssurya** [comment]
@maiqueb : can you please edit e2e's to remove the explicit exclude and see what happens? now on both l3/l2 we shouldn't have to exclude .1 and .2 anymore..

**Review by tssurya — COMMENTED** [review]
if you rebase on master you will see more e2e's skipping the gatewayIP :D

**Comment by maiqueb** [comment]
> @maiqueb : can you please edit e2e's to remove the explicit exclude and see what happens? now on both l3/l2 we shouldn't have to exclude .1 and .2 anymore..

I don't think we're actually excluding the .2 address. At least I've only seen us excluding the GW IP. 

Let me know where that's done in case I missed it. 

@tssurya

**Comment by tssurya** [comment]
> I don't think we're actually excluding the .2 address. At least I've only seen us excluding the GW IP.
> 
> Let me know where that's done in case I missed it.
> 
> @tssurya

yes you are right only .1 was excluded given we don't have a mp0 yet today.

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go [hunk 2]
```diff
@@ -31,6 +31,10 @@ func NewL2SwitchManager() *LogicalSwitchManager {
 	}
 }
 
+func NewL2SwitchManagerForUserDefinedPrimaryNetwork() *LogicalSwitchManager {
```
Comment (tssurya) [inline go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:34]: nit: func comment on why this is different from  NewL2SwitchManager and why we needed this wrapper

File: test/e2e/user-defined-primary-networks-egress.yaml
Comment (tssurya) [inline test/e2e/user-defined-primary-networks-egress.yaml:1]: oopsy

**Review by tssurya — CHANGES_REQUESTED** [review]
logic looks fine will come back for deeper round2

File: test/e2e/network_segmentation.go [hunk 3]
```diff
@@ -10,7 +10,6 @@ import (
 	nadapi "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
 	nadclient "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned/typed/k8s.cni.cncf.io/v1"
 	. "github.com/onsi/ginkgo/v2"
-	. "github.com/onsi/gomega"
```
Comment (maiqueb) [inline test/e2e/network_segmentation.go:13]: waat

File: test/e2e/user-defined-primary-networks-egress.yaml
Comment (maiqueb) [reply test/e2e/user-defined-primary-networks-egress.yaml:1]: Done.

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go (hunk 2 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:47]: Done.

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go [hunk 4]
```diff
@@ -31,6 +31,10 @@ func NewL2SwitchManager() *LogicalSwitchManager {
...
+	return NewLogicalSwitchManager()
```
Comment (qinqon) [inline go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:48]: Let's be explicit and just do, this way is easier to compare with the `NewL2SwitchManager`.
```golang
return &LogicalSwitchManager{
		allocator:  subnet.NewAllocator(),
		reserveIPs: true,
	}
```

**Review by qinqon — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go (hunk 4 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:48]: that'll be flagged as duplicate code by any tool that checks that.

File: go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go [hunk 5]
```diff
@@ -17,20 +17,37 @@ type LogicalSwitchManager struct {
...
+//     management port
 func NewL2SwitchManager() *LogicalSwitchManager {
 	return &LogicalSwitchManager{
 		allocator: subnet.NewAllocator(),
 	}
 }
 
+// NewL2SwitchManagerForUserDefinedPrimaryNetwork initializes a new logical
+// switch manager for L2 primary networks.
+// A user defined primary network auto-reserves the .1 and .2 IP addresses,
```
Comment (tssurya) [inline go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go:45]: self-note .1 is used as gateway route for pods which is why this IP needs to be excluded cause we saw failures in CI before around this.

**Review by tssurya — APPROVED** [review]
nice function comments; thanks @maiqueb

**Comment by tssurya** [comment]
https://github.com/ovn-org/ovn-kubernetes/actions/runs/9858941380/job/27223412796?pr=4509
live migration lane failed.. should I be worried or its unrelated?

**Comment by maiqueb** [comment]
> https://github.com/ovn-org/ovn-kubernetes/actions/runs/9858941380/job/27223412796?pr=4509 live migration lane failed.. should I be worried or its unrelated?

I think it's fine - the other kubevirt live migration lane passed. 

@qinqon can you evaluate this ? I get really confused w/ this lane's output. We need to sync so I learn how to make sense of this lane.

**Comment by tssurya** [comment]
```
Summarizing 1 Failure:
  [FAIL] e2e egress IP validation [OVN network] Using different methods to disable a node's availability for egress Should validate the egress IP functionality against remote hosts [It] disabling egress nodes impeding GRCP health check
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/egressip.go:470
```
known control plane failure: https://github.com/ovn-org/ovn-kubernetes/actions/runs/9858941380/job/27223408598?pr=4509

**Comment by qinqon** [comment]
> https://github.com/ovn-org/ovn-kubernetes/actions/runs/9858941380/job/27223412796?pr=4509 live migration lane failed.. should I be worried or its unrelated?

This is the fix for that, let's hope we can merge it after all the UDN madness:
- https://github.com/ovn-org/ovn-kubernetes/pull/4416

**Pull request merged**